pkg/cover/canonicalizer.go

```diff
@@ -0,0 +1,130 @@
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package cover
+
+import (
+	"sort"
+
+	"github.com/google/syzkaller/pkg/host"
+)
+
+type Canonicalizer struct {
+	// Map of modules stored as module name:kernel offset.
+	modules map[string]uint32
+
+	// Contains a sorted list of the canonical module addresses.
+	moduleKeys []uint32
+}
+
+type CanonicalizerInstance struct {
+	canonical Canonicalizer
+
+	// Contains a sorted list of the instance's module addresses.
+	moduleKeys []uint32
+
+	// Contains a map of the uint32 address to the necessary offset.
+	instToCanonicalMap map[uint32]*canonicalizerModule
+	canonicalToInstMap map[uint32]*canonicalizerModule
+}
+
+// Contains the offset and final address of each module.
+type canonicalizerModule struct {
+	offset  int
+	endAddr uint32
+}
+
+func NewCanonicalizer(modules []host.KernelModule) *Canonicalizer {
+	// Create a map of canonical module offsets by name.
+	canonicalModules := make(map[string]uint32)
+	for _, module := range modules {
+		canonicalModules[module.Name] = uint32(module.Addr)
+	}
+
+	// Store sorted canonical address keys.
+	canonicalModuleKeys := make([]uint32, len(modules))
+	setModuleKeys(canonicalModuleKeys, modules)
+	return &Canonicalizer{
+		modules:    canonicalModules,
+		moduleKeys: canonicalModuleKeys,
+	}
+}
+
+func (can *Canonicalizer) NewInstance(modules []host.KernelModule) *CanonicalizerInstance {
+	// Save sorted list of module offsets.
+	moduleKeys := make([]uint32, len(modules))
+	setModuleKeys(moduleKeys, modules)
+
+	// Create a hash between the "canonical" module addresses and each VM instance.
+	instToCanonicalMap := make(map[uint32]*canonicalizerModule)
+	canonicalToInstMap := make(map[uint32]*canonicalizerModule)
+	for _, module := range modules {
+		canonicalAddr := can.modules[module.Name]
+		instAddr := uint32(module.Addr)
+
+		canonicalModule := &canonicalizerModule{
+			offset:  int(instAddr) - int(canonicalAddr),
+			endAddr: uint32(module.Size) + canonicalAddr,
+		}
+		canonicalToInstMap[canonicalAddr] = canonicalModule
+
+		instModule := &canonicalizerModule{
+			offset:  int(canonicalAddr) - int(instAddr),
+			endAddr: uint32(module.Size) + instAddr,
+		}
+		instToCanonicalMap[instAddr] = instModule
+	}
+
+	return &CanonicalizerInstance{
+		canonical:          *can,
+		moduleKeys:         moduleKeys,
+		instToCanonicalMap: instToCanonicalMap,
+		canonicalToInstMap: canonicalToInstMap,
+	}
+}
+
+func (ci *CanonicalizerInstance) Canonicalize(cov []uint32) {
+	convertModulePCs(ci.moduleKeys, ci.instToCanonicalMap, cov)
+}
+
+func (ci *CanonicalizerInstance) Decanonicalize(cov []uint32) {
+	convertModulePCs(ci.canonical.moduleKeys, ci.canonicalToInstMap, cov)
+}
+
+// Store sorted list of addresses. Used to binary search when converting PCs.
+func setModuleKeys(moduleKeys []uint32, modules []host.KernelModule) {
+	// Determine the address that defines the end of the modules.
```

Comment:
Is `module.Addr` not actually the start of the module?

Reply:
Oh sorry, it is! I'd written this comment before deciding where to place the `module.endAddr` code. I'll delete the comment.